By Ishaq H Alsomali<h1>Ebay Car Sales Analysis</h1>

## Objective:

In this project, we attempt to clean and analyze a data set of car listings on ebay. The was shared on Kaggle but the one we are working with is not as clean as the one one Kaggle. 

In [1]:
# importing libraries
import numpy as np 
import pandas as pd 

In [2]:
# loading the data set to a panda's dataframe
autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [3]:
# display the data frame
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

## Initial Data Exploring and Cleaning

Now that we have an idea about the data frame we about to worke with, now we are interested about general information about the data set. 
- number of columns 
- data types of the columns
- any null or empty entries in a column
- storge used by the data frame

In [4]:
# basic data frame information
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

From above we have obtained the following:
- total of 20 columns.
- 50,000 total entries.
- five columns are integers and 15 are string data types.
- five columns with null or empty values.
- memory usage 4.8 MB which can be handeld on our local machine with no problems.

In [5]:
# display first 5 rows
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Exploring column names we have noticed:
- column names are ambiguous.
- column names do not follow snake case convinsion.

In [6]:
# rename columns
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']

We renamed all columns with ambiguous names and changed camel case naming style to more python convinsion (snake case)
with lower case letter and underscorer between diffrent words. 

Next, we explore more in depth different columns for describtive statistics and try identify any unormal patterns present in each column then steps can be taken after to resolve issues.

In [7]:
# describtive statistics summery
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-21 16:37:21  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

**Clearly from above we have columns with the following:**
- columns with wrong data types (price, odometer) as 'string'.
- columns containing similar values (seller,offer_type) 49,999 number of entries of 50,000 as one value.
- a column with unknown values (num_photos).

In [8]:
# analyze unique values of column num_photos
autos['num_photos'].value_counts()

0    50000
Name: num_photos, dtype: int64

We can see from above that the column (num_photos) is also dominated by one value like (price) and (odometer_km).
One thing we can do is to keep those columns in our data set and continue with analysis but since they have no added value we will drop them for the data set. 

In [9]:
# drop columns from data frame
autos = autos.drop(['seller','offer_type','num_photos'], axis=1)

In [10]:
# display changes
autos.head(1)

date_crawled                              name   price   abtest  \
0  2016-03-26 17:47:46  Peugeot_807_160_NAVTECH_ON_BOARD  $5,000  control   

  vehicle_type  registration_year  gearbox  power_ps   model   odometer  \
0          bus               2004  manuell       158  andere  150,000km   

   registration_month fuel_type    brand unrepaired_damage  \
0                   3       lpg  peugeot              nein   

            ad_created  postal_code            last_seen  
0  2016-03-26 00:00:00        79588  2016-04-06 06:45:54

- The three columns are droped from the data set.

Now we need to assign (odometer) and (price) to correct data types. 
First, we will explore each column to identify the best way to convert values. 

In [11]:
# unique values in 'odometer' column
autos['odometer'].unique()

array(['150,000km', '70,000km', '50,000km', '80,000km', '10,000km',
       '30,000km', '125,000km', '90,000km', '20,000km', '60,000km',
       '5,000km', '100,000km', '40,000km'], dtype=object)

We will remove special characters from each value in (odometer) and then convert to integer type. 
We will rename the column to (odometer_km) to preserve the meanning.

In [12]:
# clean 'odometer' then cast to integer type data
autos['odometer'] = autos['odometer'].str.replace(',','').str.replace('km','').astype(int)

In [13]:
# change column name
autos.rename({'odometer':'odometer_km'}, axis=1, inplace=True)

- Above steps will be applied to (price) column without renaming the column

In [14]:
# unique values in column (price)
autos['price'].unique()

array(['$5,000', '$8,500', '$8,990', ..., '$385', '$22,200', '$16,995'],
      dtype=object)

In [15]:
# clean 'price' then cast to float type data
autos['price'] = autos['price'].str.replace('$', '').str.replace(',','').astype(float)

## Exploring Odometer and Price

In [16]:
autos['odometer_km'].unique().shape

(13,)

In [17]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [18]:
# distripution of 'odometer_km' column
autos['odometer_km'].value_counts().sort_index(ascending=False)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64

**Findings on (odometer_km) column:**
- mean = 125732
- median = 150000
- maximum = 150000
- minimum = 5000
- 32424 cars have the maximum value 64%
- most cars are above 50000 
- it seams km values are rounded to neariest thousand

In [19]:
autos['price'].unique().shape

(2357,)

In [20]:
# discribtive stat of 'price' column
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [21]:
# distribution of 'price' column
autos['price'].value_counts().sort_index(ascending=False).head(20)

99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
259000.0      1
250000.0      1
220000.0      1
198000.0      1
197000.0      1
Name: price, dtype: int64

In [22]:
# lowest prices in the distribution
autos['price'].value_counts().sort_index(ascending=True).head(15)

0.0     1421
1.0      156
2.0        3
3.0        1
5.0        2
8.0        1
9.0        1
10.0       7
11.0       2
12.0       3
13.0       2
14.0       1
15.0       2
17.0       3
18.0       1
Name: price, dtype: int64

**Findings on (price) column:**
- mean = 9840
- median = 2950
- maximum = 100 million
- minimum = 0
- some cares has prices over 1 million USD which is't correct value
- minimum value of 0 USD is realistic since ebay is an auction site, bids might start from 0 USD.
- Prices increased dramaticaly after 350 thousand USD to the millions of USD. 

Considering any price over 350,000 USD as outliers, we will romove rows with those amounts. 
Also, considering the opening bid should start from at least  USD, we will remove 0's. 

In [23]:
# delete rows our of range (1 to 351000)
autos = autos[autos['price'].between(1,351000)]
autos['price'].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

## Exploring date data

**In the data set we have five columns as date data:**
- date_crawled 
- last_seen
- ad_created
- registration_month
- registration_year

In [24]:
# explore string type columns
autos.describe(include='object')

date_crawled         name abtest vehicle_type  gearbox  model  \
count                 48565        48565  48565        43979    46222  46107   
unique                46882        37470      2            8        2    245   
top     2016-03-30 19:48:02  Ford_Fiesta   test    limousine  manuell   golf   
freq                      3           76  25019        12598    36102   3900   

       fuel_type       brand unrepaired_damage           ad_created  \
count      44535       48565             39464                48565   
unique         7          40                 2                   76   
top       benzin  volkswagen              nein  2016-04-03 00:00:00   
freq       29368       10336             34775                 1887   

                  last_seen  
count                 48565  
unique                38474  
top     2016-04-07 06:17:27  
freq                      8

In [25]:
# display first 5 rows of 'date_crawled', 'last_seen', 'as_created' columns
autos[['date_crawled', 'last_seen', 'ad_created']].head()

date_crawled            last_seen           ad_created
0  2016-03-26 17:47:46  2016-04-06 06:45:54  2016-03-26 00:00:00
1  2016-04-04 13:38:56  2016-04-06 14:45:08  2016-04-04 00:00:00
2  2016-03-26 18:57:24  2016-04-06 20:15:37  2016-03-26 00:00:00
3  2016-03-12 16:58:10  2016-03-15 03:16:28  2016-03-12 00:00:00
4  2016-04-01 14:38:50  2016-04-01 14:38:50  2016-04-01 00:00:00

Due to the columns (date_crawled,last_seen,ad_created) are all in string formate, we can extract the first 10 characters representing actual dates for further anaylsis. 

In [26]:
# extracting only dates from the values 
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index() # normalize as percentage

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

**For the column (date_crawled):**
- the cars in the data set were crawled every day for a month period
- there are silght changes in number of cars crawled.

In [27]:
# sort distribution by date ascending order
autos['last_seen'].str[:10].value_counts(normalize=True,dropna=False).sort_index() 

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

In [28]:
# sort distribution by values decending order
autos['last_seen'].str[:10].value_counts(normalize=True,dropna=False).sort_values(ascending=False)

2016-04-06    0.221806
2016-04-07    0.131947
2016-04-05    0.124761
2016-03-17    0.028086
2016-04-03    0.025203
2016-04-02    0.024915
2016-03-30    0.024771
2016-04-04    0.024483
2016-03-12    0.023783
2016-03-31    0.023783
2016-04-01    0.022794
2016-03-29    0.022341
2016-03-22    0.021373
2016-03-28    0.020859
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-23    0.018532
2016-03-26    0.016802
2016-03-16    0.016452
2016-03-15    0.015876
2016-03-19    0.015834
2016-03-27    0.015649
2016-03-14    0.012602
2016-03-11    0.012375
2016-03-10    0.010666
2016-03-09    0.009595
2016-03-13    0.008895
2016-03-08    0.007413
2016-03-18    0.007351
2016-03-07    0.005395
2016-03-06    0.004324
2016-03-05    0.001071
Name: last_seen, dtype: float64

**For column (last_seen):**
- the distribution across last_seen seems roughly uniform except towardes last three days
- the three end days seemed to have larger percentage of cars about 10-20X more
- large percentage on last three days might not mean there was a spike in sales
- because crawling was daily, last three days might mean an increase in car listings or more entries crawled not sales
- due to fact that cars listed on ebay might be sold we will expect changes over time

In [29]:
autos['ad_created'].str[:10].value_counts(normalize=True,dropna=False).sort_index().head(20)

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
Name: ad_created, dtype: float64

In [30]:
autos['ad_created'].str[:10].value_counts(normalize=True,dropna=False).sort_values(ascending=False).head(30)

2016-04-03    0.038855
2016-03-20    0.037949
2016-03-21    0.037579
2016-04-04    0.036858
2016-03-12    0.036755
2016-03-14    0.035190
2016-04-02    0.035149
2016-03-28    0.034984
2016-03-07    0.034737
2016-03-29    0.034037
2016-03-15    0.034016
2016-03-19    0.033687
2016-04-01    0.033687
2016-03-30    0.033501
2016-03-08    0.033316
2016-03-09    0.033151
2016-03-11    0.032904
2016-03-22    0.032801
2016-03-26    0.032266
2016-03-23    0.032060
2016-03-10    0.031895
2016-03-31    0.031875
2016-03-25    0.031751
2016-03-17    0.031278
2016-03-27    0.030989
2016-03-16    0.030125
2016-03-24    0.029280
2016-03-05    0.022897
2016-03-13    0.017008
2016-03-06    0.015320
Name: ad_created, dtype: float64

**From column(ad_created):**
- majority of cars were listed on same months the data set was crawled
- number of cars were listed up to 9 months prior crawling

In [31]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

In [32]:
# unique values in 'registration_year' column
autos['registration_year'].unique()

array([2004, 1997, 2009, 2007, 2003, 2006, 1995, 1998, 2000, 2017, 2010,
       1999, 1982, 1990, 2015, 2014, 1996, 1992, 2002, 2012, 2011, 2005,
       2008, 1985, 2016, 1994, 1986, 2001, 2018, 2013, 1972, 1993, 1988,
       1989, 1973, 1967, 1976, 4500, 1987, 1991, 1983, 1960, 1969, 1950,
       1978, 1980, 1984, 1963, 1977, 1961, 1968, 1934, 1965, 1971, 1966,
       1979, 1981, 1970, 1974, 1910, 1975, 5000, 4100, 2019, 1956, 9999,
       6200, 1964, 1959, 1958, 1800, 1948, 1931, 1943, 1941, 1962, 1927,
       1937, 1929, 1000, 1957, 1952, 1111, 1955, 1939, 8888, 1954, 1938,
       2800, 5911, 1953, 1951, 4800, 1001, 9000], dtype=int64)

We observed a wrong data in the column(registration_year) where 
- maximum year is 9999 which is not a year
- minimum year is 1000 which is not a year

Findings suggests wrong entries in the data set which needs more investigation. 

In [33]:
# registration_year values out of range
autos[~autos['registration_year'].between(1900,2016)]['registration_year'].value_counts(dropna=False)

2017    1392
2018     470
5000       4
9999       3
2019       2
1800       2
2800       1
6200       1
4500       1
8888       1
4800       1
1000       1
9000       1
1001       1
1111       1
5911       1
4100       1
Name: registration_year, dtype: int64

In [34]:
# number of out of range values in the column 'registration_year'
autos[~autos['registration_year'].between(1900,2016)]['registration_year'].value_counts(dropna=False).sum()

1884

We have 1884 cars with registration dates beyound reasanble range of dates.
Our choice of limits is based on that cars before 1900 are not as common as well as cars made beyound 2016 the year the data set was obtained are not useful in this analysis. 

In [35]:
# remove rows for out of range values
autos = autos[autos['registration_year'].between(1900,2016)]

In [36]:
autos['registration_year'].describe()

count    46681.000000
mean      2002.910756
std          7.185103
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

In [37]:
autos['registration_year'].value_counts(normalize=True,dropna=False).sort_index(ascending=False)

2016    0.026135
2015    0.008397
2014    0.014203
2013    0.017202
2012    0.028063
          ...   
1934    0.000043
1931    0.000021
1929    0.000021
1927    0.000021
1910    0.000107
Name: registration_year, Length: 78, dtype: float64

In [38]:
autos['registration_year'].value_counts(normalize=True,dropna=False).sort_values(ascending=False)

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
          ...   
1931    0.000021
1929    0.000021
1943    0.000021
1953    0.000021
1952    0.000021
Name: registration_year, Length: 78, dtype: float64

**From column(registration_year):**
- We had entries with wrong registration dates and we had to remove them from the set
- it seems more cars in the data set with registration year from 1990 to 2016
- less cars are registered in years before 1970 getting a lot rare for years earlier than 1950
- earliest registration date maps to 1910
- latest car registered on 2016

## Exploring prices of top 5 brands

We are interested on analyzing the prices of top 5 brands in the data set. 

In [39]:
autos['brand'].value_counts(dropna=False).sort_values(ascending=False)

volkswagen        9862
bmw               5137
opel              5022
mercedes_benz     4503
audi              4041
ford              3263
renault           2201
peugeot           1393
fiat              1197
seat               853
skoda              766
nissan             713
mazda              709
smart              661
citroen            654
toyota             593
hyundai            468
sonstige_autos     458
volvo              427
mini               409
mitsubishi         384
honda              366
kia                330
alfa_romeo         310
porsche            286
suzuki             277
chevrolet          266
chrysler           164
dacia              123
daihatsu           117
jeep               106
subaru             100
land_rover          98
saab                77
jaguar              73
daewoo              70
trabant             65
rover               62
lancia              50
lada                27
Name: brand, dtype: int64

- It seems top 5 brands in terms of number of listing is domenated by german brands.

In [40]:
# selecting top 5 brands 
top_5_brands = autos['brand'].value_counts().sort_values(ascending=False).head()

In [41]:
print('top 5 brands percentage of all listing:',(top_5_brands.sum()/autos['brand'].value_counts().sum())*100)

top 5 brands percentage of all listing: 61.191919624686705


Now, we would like to find the average price of the top 5 cars in the data sets. 

In [42]:
# list of top 5 brands names
brands_list = top_5_brands.index

In [43]:
# for loop to calculate average price by top 5 brands
brand_price_avg = {} 
for brand in brands_list: 
    mean = autos[autos['brand']==brand]['price'].mean()
    brand_price_avg[brand] =round(mean)
    
print('average price of all cars in data set:', round(autos['price'].mean()))
print('------------------------------------')
brand_price_avg

average price of all cars in data set: 5978.0
------------------------------------


{'volkswagen': 5402.0,
 'bmw': 8333.0,
 'opel': 2975.0,
 'mercedes_benz': 8628.0,
 'audi': 9337.0}

**Findings:**
- Top 5 brands in the data set are domenated by german brands
- The average prices of top 5 brands varies as audi is the largest and opel with the lowest
- 3 out of 5 brands are above the total proce average for all brands
- on average audi, mercedes_benz and bmw are the most expensive 
- opel is the cheapest on average and volkswagen is in between

Now, we would like to find the average kilometers of the top 5 cars in the data sets. 

In [44]:
# for loop to calculate average distance by top 5 brands
brand_distance_avg = {}
for brand in brands_list:
    mean = autos[autos['brand']==brand]['odometer_km'].mean()
    brand_distance_avg[brand] = round(mean)
    
print('average distance of all cars in data set:', round(autos['odometer_km'].mean()))
print('------------------------------------')
brand_distance_avg

average distance of all cars in data set: 125587.0
------------------------------------


{'volkswagen': 128707.0,
 'bmw': 132573.0,
 'opel': 129310.0,
 'mercedes_benz': 130788.0,
 'audi': 129157.0}

**Findings:**
- all brands had total average of distance covered of 125,587 kilometers
- top 5 brands had average distance (kilometers) higher than than the total average kilometers
- it seems the distance covered by top 5 brands are somewhat similar with minner difference
- average kilometers is not decisive factor explaning the price difference between the brands

In [45]:
# construct a series from dictionary
brand_price_avg = pd.Series(brand_price_avg) 
brand_distance_avg = pd.Series(brand_distance_avg)
# construct data frame from series object 
df_brands = pd.DataFrame(brand_price_avg, columns=['mean_price'])
df_brands['mean_km'] = brand_distance_avg
df_brands

mean_price   mean_km
volkswagen         5402.0  128707.0
bmw                8333.0  132573.0
opel               2975.0  129310.0
mercedes_benz      8628.0  130788.0
audi               9337.0  129157.0

**Results:**
- of all brands, volkswagen is the most popular having more listings than mercedes_benz and audi combined
- Germen brands are the most popular brands in the data set accounting for nearly 61% of all listings
- audi, mercedes_benz and bmw have higher price average than volkswagen and opel
- all top 5 brands have on average higher distance covered than total distance average of all brands